File: UnivercityProject/src/main/java/com/example/UnivercityProject/admin/service/AdminService.java
package com.example.UnivercityProject.admin.service;

import com.example.UnivercityProject.admin.repo.AdminRepo;
import com.example.UnivercityProject.choice.entity.Choice;
import com.example.UnivercityProject.choice.repo.ChoiceRepo;
import com.example.UnivercityProject.classroom.entity.Classroom;
import com.example.UnivercityProject.classroom.repo.ClassroomRepo;
import com.example.UnivercityProject.exam.entity.Exam;
import com.example.UnivercityProject.exam.repo.ExamRepo;
import com.example.UnivercityProject.question.entity.Question;
import com.example.UnivercityProject.question.repo.QuestionRepo;
import com.example.UnivercityProject.result.entity.Result;
import com.example.UnivercityProject.result.repo.ResultRepo;
import com.example.UnivercityProject.student.entity.Student;
import com.example.UnivercityProject.student.repo.StudentRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class AdminService {

    @Autowired
    private ClassroomRepo classroomRepo;
    @Autowired
    private StudentRepo studentRepo;
    @Autowired
    private ExamRepo examRepo;
    @Autowired
    private QuestionRepo questionRepo;
    @Autowired
    private ChoiceRepo choiceRepo;
    @Autowired
    private AdminRepo adminRepo;
    @Autowired
    private ResultRepo resultRepo;

    // Classroom CRUD
    public Classroom createClassroom(Classroom classroom) {
        return classroomRepo.save(classroom);
    }

    public List<Classroom> getAllClassrooms() {
        return classroomRepo.findAll();
    }

    public Classroom getClassroomById(String classroomId) {
        return classroomRepo.findById(classroomId).orElse(null);
    }

    public Classroom updateClassroom(String classroomId, String updatedClassroomName) {
        Classroom classroom = classroomRepo.findById(classroomId).orElse(null);
        if (classroom != null) {
            classroom.setName(updatedClassroomName);
            return classroomRepo.save(classroom);
        }
        return null;
    }

    public void deleteClassroomWithRelatedEntities(Classroom classroom) {
        // Classroom nesnesine ait tüm examları bul
        List<Exam> exams = examRepo.findByClassroom(classroom);
        // Her exam için related questionları bul
        for (Exam exam: exams){
            List<Question> questions = questionRepo.findByExam(exam);
            // Her question için related choiceları bul ve sil ardından question nesnesini sil
            for (Question question: questions) {
                List<Choice> choices = choiceRepo.findByQuestion(question);
                choiceRepo.deleteAll(choices);
                questionRepo.delete(question);
            }
            examRepo.deleteAll(exams);
            List<Result> results = resultRepo.findByExam(exam);
            resultRepo.deleteAll(results);
        }
        List<Student> students = studentRepo.findByClassroom(classroom);
        for (Student student: students) {
            List<Result> results = resultRepo.findByStudent(student);
            resultRepo.deleteAll(results);
            studentRepo.delete(student);
        }
        classroomRepo.delete(classroom);
    }

    // Student CRUD
    public Student createStudent(Student student) {
        return studentRepo.save(student);
    }

    public List<Student> getAllStudents() {
        return studentRepo.findAll();
    }

    public Student getStudentById(String studentId) {
        return studentRepo.findById(studentId).orElse(null);
    }

    public Student updateStudent(String studentId, Student updatedStudent) {
        Student student = studentRepo.findById(studentId).orElse(null);
        if (student != null) {
            student.setName(updatedStudent.getName());
            student.setNumber(updatedStudent.getNumber());
            student.setPassword(updatedStudent.getPassword());
            student.setClassroom(updatedStudent.getClassroom());
            return studentRepo.save(student);
        }
        return null;
    }

    public void deleteStudentWithRelatedEntities(Student student) {
        List<Result> results = resultRepo.findByStudent(student);
        for (Result result: results) {
            resultRepo.delete(result);
        }
        studentRepo.delete(student);
    }

    // Exam CRUD
    public Exam createExam(Exam exam) {
        return examRepo.save(exam);
    }

    public List<Exam> getAllExams() {
        return examRepo.findAll();
    }

    public Exam getExamById(String examId) {
        return examRepo.findById(examId).orElse(null);
    }

    public Exam updateExam(String examId, Exam updatedExam) {
        Exam exam = examRepo.findById(examId).orElse(null);
        if (exam != null) {
            exam.setName(updatedExam.getName());
            exam.setStartDate(updatedExam.getStartDate());
            exam.setEndDate(updatedExam.getEndDate());
            exam.setDuration(updatedExam.getDuration());
            exam.setClassroom(updatedExam.getClassroom());
            return examRepo.save(exam);
        }
        return null;
    }

    public void deleteExamWithRelatedEntities(Exam exam) {
        // Exam nesnesine ait tüm questionları bul
        List<Question> questions = questionRepo.findByExam(exam);

        // Her question için related choice'ları bul ve sil ardından question nesnesini sil
        for (Question question : questions) {
            List<Choice> choices = choiceRepo.findByQuestion(question);
            choiceRepo.deleteAll(choices);
            questionRepo.delete(question);
        }

        // Exam nesnesine ait tüm resultları bul ve sil
        List<Result> results = resultRepo.findByExam(exam);
        resultRepo.deleteAll(results);

        // Exam nesnesini sil
        examRepo.delete(exam);
    }

    // Question CRUD
    public Question createQuestion(Question question) {
        return questionRepo.save(question);
    }

    public List<Question> getAllQuestions() {
        return questionRepo.findAll();
    }

    public Question getQuestionById(String questionId) {
        return questionRepo.findById(questionId).orElse(null);
    }

    public Question updateQuestion(String questionId, Question updatedQuestion) {
        Question question = questionRepo.findById(questionId).orElse(null);
        if (question != null) {
            question.setTitle(updatedQuestion.getTitle());
            question.setQuestionNumber(updatedQuestion.getQuestionNumber());
            question.setExam(updatedQuestion.getExam());
            return questionRepo.save(question);
        }
        return null;
    }

    public void deleteQuestionWithRelatedEntities(Question question) {
            List<Choice> choices = choiceRepo.findByQuestion(question);
        for (Choice choice: choices) {
            choiceRepo.delete(choice);
        }
        questionRepo.delete(question);
    }

    // Choice CRUD
    public Choice createChoice(Choice choice) {
        return choiceRepo.save(choice);
    }

    public List<Choice> getAllChoices() {
        return choiceRepo.findAll();
    }

    public Choice getChoiceById(String choiceId) {
        return choiceRepo.findById(choiceId).orElse(null);
    }

    public Choice updateChoice(String choiceId, Choice updatedChoice) {
        Choice choice = choiceRepo.findById(choiceId).orElse(null);
        if (choice != null) {
            choice.setName(updatedChoice.getName());
            choice.setDescription(updatedChoice.getDescription());
            choice.setCorrect(updatedChoice.isCorrect());
            choice.setQuestion(updatedChoice.getQuestion());
            return choiceRepo.save(choice);
        }
        return null;
    }

    public void deleteChoiceById(String choiceId) {
        choiceRepo.deleteById(choiceId);
    }

    // Result CRUD
    public Result createResult(Result result) {return resultRepo.save(result);}

    public List<Result> getAllResults() { return resultRepo.findAll();}

    public Result getResultById(String resultId) {
        return resultRepo.findById(resultId).orElse(null);
    }

    public Result updateResult(String resultId, Result updatedResult) {
        Result result = resultRepo.findById(resultId).orElse(null);
        if (result != null) {
            result.setScore(updatedResult.getScore());
            result.setStudent(updatedResult.getStudent());
            result.setExam(updatedResult.getExam());
            return resultRepo.save(result);
        }
        return null;
    }

    public void deleteResultById(String resultId) {
        resultRepo.deleteById(resultId);
    }
}

File: UniFrontend/src/app/services/admin/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Classroom } from '../../model/classroom.model';
import { Student } from 'src/app/model/student.model';
import { Exam } from 'src/app/model/exam.model';
import { Question } from 'src/app/model/question.model';
import { Choice } from 'src/app/model/choice.model';
import { Result } from 'src/app/model/result.model';

@Injectable({
  providedIn: 'root'
})
export class AdminService {

  private baseUrl = 'http://localhost:8080/admin'; 

  constructor(private http: HttpClient) { }

  // Classroom CRUD
  createClassroom(classroom: Classroom): Observable<Classroom> {
    return this.http.post<Classroom>(`${this.baseUrl}/classrooms`, classroom);
  }
  getClassrooms(): Observable<Classroom[]> {
    return this.http.get<Classroom[]>(`${this.baseUrl}/classrooms`);
  }
  updateClassroom(classroom: Classroom): Observable<Classroom> {
    return this.http.put<Classroom>(`${this.baseUrl}/classrooms/${classroom.id}`, classroom);
  }
  deleteClassroom(classroom: Classroom): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/classrooms/${classroom.id}`);
  }
  // Student CRUD
  createStudent(student: Student): Observable<Student> {
    return this.http.post<Student>(`${this.baseUrl}/students`, student);
  }
  getAllStudents(): Observable<Student[]> {
    return this.http.get<Student[]>(`${this.baseUrl}/students`);
  }
  getStudentById(studentId: string): Observable<Student> {
    return this.http.get<Student>(`${this.baseUrl}/students/${studentId}`);
  }
  updateStudent(student: Student): Observable<Student> {
    return this.http.put<Student>(`${this.baseUrl}/students/${student.id}`,student);
  }
  deleteStudentWithRelatedEntities(student: Student): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/students/${student.id}`);
  }
  // Exam CRUD 
  createExam(exam: Exam): Observable<Exam> {
    return this.http.post<Exam>(`${this.baseUrl}/exams`, exam);
  }
  getAllExams(): Observable<Exam[]> {
    return this.http.get<Exam[]>(`${this.baseUrl}/exams`);
  }
  getExamById(examId: string): Observable<Exam> {
    return this.http.get<Exam>(`${this.baseUrl}/exams/${examId}`);
  }
  updateExam(exam: Exam): Observable<Exam> {
    return this.http.put<Exam>(`${this.baseUrl}/exams/${exam.id}`, exam);
  }
  deleteExamWithRelatedEntities(exam: Exam): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/exams/${exam.id}`);
  }
  // Question CRUD
  createQuestion(question: Question): Observable<Question> {
    return this.http.post<Question>(`${this.baseUrl}/questions`, question);
  }
  getAllQuestions(): Observable<Question[]> {
    return this.http.get<Question[]>(`${this.baseUrl}/questions`);
  }
  getQuestionById(questionId: string): Observable<Question> {
    return this.http.get<Question>(`${this.baseUrl}/questions/${questionId}`);
  }
  updateQuestion(question: Question): Observable<Question> {
    return this.http.put<Question>(`${this.baseUrl}/questions/${question.id}`, question);
  }
  deleteQuestion(question: Question): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/questions/${question.id}`);
  }
  // Choice CRUD
  createChoice(choice: Choice): Observable<Choice> {  
    return this.http.post<Choice>(`${this.baseUrl}/choices`, choice);
  }
  getAllChoices(): Observable<Choice[]> {
    return this.http.get<Choice[]>(`${this.baseUrl}/choices`);
  }
  getChoiceById(choiceId: string): Observable<Choice> {
    return this.http.get<Choice>(`${this.baseUrl}/choices/${choiceId}`);
  }
  updateChoice(choice: Choice): Observable<Choice> {
    return this.http.put<Choice>(`${this.baseUrl}/choices/${choice.id}`, choice);
  }
  deleteChoice(choice: Choice): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/choices/${choice.id}`);
  }
  // Result CRUD
  createResult(result: Result): Observable<Result> {  
    return this.http.post<Result>(`${this.baseUrl}/results`, result);
  }
  getAllResults(): Observable<Result[]> {
    return this.http.get<Result[]>(`${this.baseUrl}/results`);
  }
  getResultById(resultId: string): Observable<Result> {
    return this.http.get<Result>(`${this.baseUrl}/results/${resultId}`);
  }
  updateResult(result: Result): Observable<Result> {
    return this.http.put<Result>(`${this.baseUrl}/results/${result.id}`, result);
  }
  deleteResult(result: Result): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/results/${result.id}`);
  }
  getStudentScore(studentId: string, examId: string): Observable<number> {
    return this.http.get<number>(`http://localhost:8080/result/student-score?studentId=${studentId}&examId=${examId}`);
  }
  
}
